Move wpilibj and wpiutil.math inside wpimath to edu.wpi.first.math

noamzaks opened the pull request:
Moved:

wpimath/src/main/java/edu/wpi/first/wpiutil/math --> wpimath/src/main/java/edu/wpi/first/math
wpimath/src/main/java/edu/wpi/first/wpilibj --> wpimath/src/main/java/edu/wpi/first/math
Refactored JavaDoc and imports (with help from IntelliJ).

I didn't move edu.first.wpi.wpilibj.controller.

`gradlew test` built succesfully for me on macOS.

auscompgeek commented on wpilibj/src/test/java/edu/wpi/first/wpilibj/SlewRateLimiterTest.java:
```diff
@@ -7,6 +7,7 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+import edu.wpi.first.math.filter.SlewRateLimiter;
```
Hmm. Seems like this unit test belongs in wpimath too.

noamzaks commented:
gradlew build was successful this time on my Mac

PeterJohnson commented:
> I didn't move edu.first.wpi.wpilibj.controller.

Is there a reason why you didn't?  I would think we should go ahead and move that as well.

noamzaks commented:
> Is there a reason why you didn't?  I would think we should go ahead and move that as well.

There isn't, other than I wasn't sure if both that module inside wpilibj and that module inside wpimath are related to math. 

My qurstion is: do you want to move both wpilibj/src/main/java/edu/wpi/first/wpilibj/controller and wpimath/src/main/java/edu/wpi/first/wpilibj/controller?
Or only the one inside wpimath?

PeterJohnson commented:
@calcmogul thoughts here?  Should the wpilibj controller classes actually be moving to wpimath?

calcmogul commented:
Yes, we should move all controllers to wpimath that we can.

noamzaks commented:
> Yes, we should move all controllers to wpimath that we can.

From the controllers which are inside wpilibj sources, I think only the RamseteController can move entirely into wpimath (because the others have a dependency on wpilibj).
Do you want to move it also? I've moved all of the controller classes into the math package, but inside wpilibj currently.

calcmogul commented:
wpilibj.controllers.PIDController should be able to move.

noamzaks commented:
Maybe we can try to decouple the Sendable and HAL part from the controllers and use something like a (module-private?) MathPIDController of some sorts, then inheriting it inside wpilibj and just adding the triggers?

PeterJohnson commented:
Not package-private, as wpilibj and wpimath are in different packages, but yes, that could work.  We would need to think about how to handle C++ in that approach though (this PR right now is only addressing Java).

noamzaks commented:
We can do the same thing with C++?

PeterJohnson commented:
Only if we name it uniquely.  Right now wpimath is in the same namespace as wpilibc, so we'd have to rename the class--in Java we could actually keep the class the same name as long as it was in a different package (although admittedly a unique class name is probably the right thing to do).

calcmogul commented:
Given these two choices:

1. `wpimath.PIDControllerImpl` and `wpilibj.PIDController implements Sendable` that uses `wpimath.PIDControllerImpl`
2. `wpilibj.PIDController implements Sendable`

I'd rather have (2).

noamzaks commented:
Keep things they way they are?

calcmogul commented:
Yea. In my opinion, the architectural gymnastics of splitting them doesn't seem worth it.

noamzaks commented:
That makes sense, tho I can also see why you’d want to put the math separately from wpilibj/wpilibc to make sure the math “responsibility” is in wpimath

PeterJohnson commented:
Oh, you moved the controllers and filters in wpilibj to the math package too.  I'm not sure we want to do that vs. just leaving it where it is... we aren't using java modules yet but doing things like that could be problematic down the road.  @calcmogul ?

calcmogul commented:
Problematic in what way?

PeterJohnson commented:
@noamzaks what we want to do is put everything in wpimath in edu.wpi.first.math.xx and everything in wpilibj in edu.wpi.first.wpilibj.xx.  So you need to undo the change you made in wpilibj.

noamzaks commented:
Just to confirm, you wish to move the wpilibj-part of the controller package from math back to wpilibj? (wpilibj/.../math/controller$)
Keep the math-part of the controller package within math? (wpimath/.../math/controller)
Also, what about SlewRateLimiter? (wpilibj/.../math/filter or wpilibj/.../wpilibj)

PeterJohnson commented:
If SlewRateLimiter doesn’t depend on Sendable, it can be moved to wpimath and put into edu.wpi.first.math.filter

noamzaks commented:
SlewRateLimiter depends on `Timer`

prateekma commented:
If it only depends on `Timer.getFPGATimestamp()`, you can replace it with `WPIUtilJNI.now()` (and check units).

noamzaks commented:
I moved SlewRateLimiter and switched to WPIUtilJNI.now().

- The test was using Timer.delay(), so I am unsure how to test the current implementation (I tried Thread.sleep the same way it is called in Timer.delay(), but it failed). The test is currently commented out, so I am converting this to a draft for now.
- I traced WPIUtil.now() through the WPIUtilJNI.cpp to timestamp.h where the units are microseconds. This should probably be verified and tested, and maybe a private class function to return the current time in seconds can clean up the way it looks now.

PeterJohnson commented:
It's possible to mock time using the wpiutil functions, but not from Java.  Let's move the SlewRateLimiter change to a different PR so we can merge the rest of this one.
